File: apps/android-client/src/NavigationHost.kt
package com.gogate.android

class NavigationHost {
    private val allowed = setOf("home", "server_list", "connection_details", "logs", "settings", "about")

    var currentRoute: String = "home"
        private set

    private val listeners = mutableListOf<(String) -> Unit>()

    fun navigate(route: String): Boolean {
        val normalized = route.trim().lowercase()
        if (!allowed.contains(normalized)) return false
        if (currentRoute == normalized) return true
        currentRoute = normalized
        listeners.forEach { it(currentRoute) }
        return true
    }

    fun subscribe(listener: (String) -> Unit): DisposableSubscription {
        listeners += listener
        return DisposableSubscription { listeners.remove(listener) }
    }
}

File: apps/windows-client/src/NavigationHost.cs
using System;
using System.Collections.Generic;

namespace GoGate.WindowsClient;

public class NavigationHost
{
    private static readonly HashSet<string> AllowedRoutes = new(StringComparer.OrdinalIgnoreCase)
    {
        "home", "settings", "ip_scanner", "logs", "about"
    };

    public string CurrentRoute { get; private set; } = "home";

    public event Action<string>? OnRouteChanged;

    public bool Navigate(string route)
    {
        var normalized = route?.Trim().ToLowerInvariant() ?? string.Empty;
        if (!AllowedRoutes.Contains(normalized))
            return false;
        if (CurrentRoute == normalized)
            return true;

        CurrentRoute = normalized;
        OnRouteChanged?.Invoke(CurrentRoute);
        return true;
    }
}
